modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,472 @@
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// The header is missing digests related to consensus events.
+	///
+	/// This will typically have to do with missing authority set change signals.
+	MissingConsensusDigest,
+}
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait ChainVerifier<S, H> {
+	/// Import a header to the pallet.
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError>;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	///
+	/// Returns a list of headers which got finalized by the given header.
+	fn verify_finality(storage: &mut S, header: &H, proof: FinalityProof) -> Result<Vec<H>, ImportError>;
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier;
+
+impl<S, H> ChainVerifier<S, H> for Verifier
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError> {
+		let highest_finalized = storage.best_finalized_header().expect("TODO");
+		if header.number() < highest_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// A block at this height should come with a justification and signal a new
+		// authority set. We'll want to make sure it is valid
+		//
+		// This defaults to 0, should it maybe be an Option?
+		let scheduled_change_height = storage.scheduled_set_change().height;
+		if *header.number() == scheduled_change_height && finality_proof.is_some() {
+			// Maybe pass the scheduled_change in here so we don't have to query storage later
+			let finalized_headers = Self::verify_finality(
+				storage,
+				header,
+				finality_proof.expect("Checked for `finality_proof` before entering if-block"),
+			)?;
+
+			// TODO: Would need to prune blocks from the non-canonical chain at some point
+			for header in finalized_headers.iter() {
+				storage.write_header(&ImportedHeader::new(header.clone(), true));
+			}
+
+			return Ok(());
+		}
+
+		// Don't like having to take ownership of header...
+		storage.write_header(&ImportedHeader::new(header.clone(), false));
+
+		Ok(())
+	}
+
+	fn verify_finality(storage: &mut S, header: &H, proof: FinalityProof) -> Result<Vec<H>, ImportError> {
```

Comment:
I'd prefer to not mix storage reads & writes here. Since the `import_header` function is already doing some writes, I think we should keep a really strict boundary between verification (state reads), preparation of update data (also only state reads), error handling and actual state writes. Otherwise it's really easy to miss `?` which can leave the state inconsistent.

Reply:
I was trying to keep it kinda split. Like I mentioned to Slava earlier, there are no header writes in `verify_finality`, that only happens in `import_headers`. The only things that are written during the finality verification process are the authority set changes. It might be a little strange to have the authority set changes updated in `import_header`, but  I can look into doing all the writes there.